### PR TITLE
multi-token mapping for runnables.rs (#10070)

### PR DESCRIPTION
**Pull request by spookyvision**
implement #10070 in [`runnables`](https://github.com/rust-analyzer/rust-analyzer/blob/master/crates/ide/src/runnables.rs)

### CODE REVIEW

**Review by lnicola — APPROVED** [review]

File: crates/ide/src/runnables.rs [hunk 1]
```diff
@@ -229,15 +229,24 @@ fn find_related_tests(
...
+                        vec![]
+                    }
+                }
             });
 
+            // find first suitable ancestor
+            let functions = tokens
+                .map(|token| token.ancestors().find_map(ast::Fn::cast))
+                // remove instances where no ancestor could be found ( = find_map returns `None`)
+                .filter_map(|f| f)
```
Comment (jonas-schievink) [inline crates/ide/src/runnables.rs:247]:
```suggestion
                .filter_map(|token| token.ancestors().find_map(ast::Fn::cast))
```

File: crates/ide/src/runnables.rs [hunk 2]
```diff
@@ -229,15 +229,24 @@ fn find_related_tests(
...
-                let token = sema.descend_into_macros(token);
-                token
-                    .ancestors()
-                    .find_map(ast::Fn::cast)
-                    .map(|f| hir::InFile::new(sema.hir_file_for(f.syntax()), f))
+
+            // create flattened vec of tokens
+            let tokens = refs.iter().flat_map(|(range, _)| {
+                match file.token_at_offset(range.start()).next() {
+                    Some(token) => sema.descend_into_macros_many(token).into_vec(),
```
Comment (jonas-schievink) [inline crates/ide/src/runnables.rs:236]: This returns a `SmallVec`, so you can avoid the additional `Vec` allocation

**Review by jonas-schievink — COMMENTED** [review]
Looks good! I haven't figured out a way to test this yet, since "Peek Related Tests" does not work on `first` and `second` in the following code:

```rust
fn first() {}
fn second() {}

macro_rules! m {
    ($test:ident -> $first:ident, $second:ident) => {
        #[test]
        fn $test() {
            $first();
        }

        mod module {
            #[test]
            fn $test() {
                crate::$second();
            }
        }
    };
}

m!(abc -> first, second);
```

File: crates/ide/src/runnables.rs (hunk 2 above)
Comment (spookyvision) [reply crates/ide/src/runnables.rs:236]: I could not get that to work with the `None =>` branch without introducing an explicit dependency on `SmallVec` - am I missing something?

**Comment by spookyvision** [comment]
> Looks good! I haven't figured out a way to test this yet, since "Peek Related Tests" does not work on `first` and `second` in the following code:
> 
> ```rust
> // ...
> ```

Does it depend on some other aspect of this ticket or another missing feature?

**Comment by jonas-schievink** [comment]
> Does it depend on some other aspect of this ticket or another missing feature?

I don't think it's part of #10070, but I'm not really sure what exactly the problem is either

**Comment by jonas-schievink** [comment]
bors d+

Looks like CI will do the remaining review, feel free to r+ once that passes

**Comment by lnicola** [comment]
Squashing should fix the CI error.

**Comment by lnicola** [comment]
But you didn't squash..

**Comment by spookyvision** [comment]
> But you didn't squash..

sorry - how should I fix this now?

**Comment by lnicola** [comment]
Untested:

```
$ git rebase -i @~4
# edit the file to put `p` or `pick` on the last three commits (if vim starts, you'll need to press `i` to insert, then quit with `ESC` and  `:wq`
$ git log # check that there's only one commit
$ git show # check that the commit looks correct
$ git push --force-with-lease
```

**Comment by lnicola** [comment]
Umm, that didn't come out right.

**Comment by lnicola** [comment]
Ah, sorry, that should have been `fixup`, not `pick`.

**Comment by lnicola** [comment]
Try:

```
$ git reset --hard e2b6b0a6bba3b1ef56461d01ead89a69ad17549a
$ git show # make sure it looks correct
$ git push --force-with-lease
```

**Comment by spookyvision** [comment]
> Untested:
> 
> ```
> # ...
> ```

sorry, I wasn't clear in my phrasing, I wanted to ask how to undo the bors action 😆  - that seems resolved after your `r-`?

(rebase help is also always appreciated, but I think I made it work now, after making it much worse temporarily)

**Comment by lnicola** [comment]
bors r-

I already filed and r+ed #10188.

> sorry, I wasn't clear in my phrasing, I wanted to ask how to undo the bors action laughing - that seems resolved after your r-?

Oh, sorry.

**Pull request closed without merging**